New tutorial - single cell data import and format conversion

wee-snufkin opened the pull request:
Tutorial describing the most common single-cell datatypes, how to import data using EBI SCXA and HCA tools, and how to convert between the formats. 

Can leave it as one big tutorial or split into separate smaller ones. 

@nomadscientist 

UPDATES AFTER THE REVIEW:
- [x]  move the data import section from here to the separate tutorial
- [x] use downsampled AnnData input file
- [x] move CDS data conversion section from monocle tutorial into here
- [x] update histories, workflows and tests
- [x] add Seurat -> AnnData conversion (SCEasy)

hexylena commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,561 @@
+
+However, sometimes it is useful to know how to do this conversion manually or at least to know how it all works. Therefore, below are some examples showing how to do it. 
+
+## AnnData -> Seurat
```
```suggestion
## AnnData → Seurat
```

hexylena reviewed (commented):
Wish we had more topic introduction tutorials like this, this is so fantastic!

nomadscientist commented:
Wooo! Testing this TODAY!

wee-snufkin commented:
Will try to fix the workflow tests later today...

nomadscientist commented:
I think the 'downsampling' section should go AFTER the Conversion section, as it will be less common that a user needs to do it

nomadscientist commented:
For downsampling, do you not need to also perform this on the transcripts file so that the two can then be combined?

mtekman commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,561 @@
+>  
+> The updated SCEasy tool is called [**SCEasy Converter**](toolshed.g2.bx.psu.edu/repos/iuc/sceasy_convert/sceasy_convert/0.0.7+galaxy1) and it's only available on usegalaxy.eu. The second tool is called [**SCEasy convert**](toolshed.g2.bx.psu.edu/repos/ebi-gxa/sceasy_convert/sceasy_convert/0.0.5+galaxy1) and it works on usegalaxy.org, however has limited conversion options.
+> 
+{: .warning}
```
Whoops, we actually wanted to avoid this

The idea was that the new sceasy tool in the IUC repo would be a direct update of the EBI-GXA repo (indeed, that's what I based it on). I pushed the tool to IUC because I don't have ebi access and long-term wise, iuc made more sense.

Maybe we retire the old sceasy tool @pcm32 to avoid any confusion?

pcm32 replied:
Please PR that. Anyone can PR to that repo, I think it might have been better to do the change there and avoid the change in identifier for the tool. I would appreciate if we can avoid this pattern in the future, thanks!

nomadscientist commented:
The tutorial takes a long time to run, so I wonder if you should use the downsampled or a downsampled AnnData object to do the conversions rather than the full mito-anndata?

nomadscientist commented:
I'd say that this section
AnnData --> CDS

You could just move that entire section from the Monocle tutorial to this new one.
Then you can just reference/link to it in the Monocle tutorial, and start people off with the data in the right format there.

That's nice because then courses can decide how much data manipulation vs analysis to include, AND it is consistent with the rest of the tutorials

wee-snufkin commented:
@nomadscientist your suggestions applied, good to go from my end! 

@hexylena does linting fail because of linking the subsection? What else is wrong? Lmk if/how I can fix the checks 

@mtekman I tested the AnnData object in .h5ad (successfully converted from Seurat by the new SCEasy Converter) using _Inspect AnnData_ tool, but it failed with the following error: 
`Traceback (most recent call last):
  File "/usr/local/tools/_conda/envs/mulled-v1-f1891eb62511084a03c08c08b82a6b785abdf465b18ed9ac4a36c19217d84a96/lib/python3.9/site-packages/anndata/_io/utils.py", line 156, in func_wrapper
    return func(elem, *args, **kwargs)
  File "/usr/local/tools/_conda/envs/mulled-v1-f1891eb62511084a03c08c08b82a6b785abdf465b18ed9ac4a36c19217d84a96/lib/python3.9/site-packages/anndata/_io/h5ad.py", line 532, in read_group
    EncodingVersions[encoding_type].check(
  File "/usr/local/tools/_conda/envs/mulled-v1-f1891eb62511084a03c08c08b82a6b785abdf465b18ed9ac4a36c19217d84a96/lib/python3.9/enum.py", line 408, in __getitem__
    return cls._member_map_[name]
KeyError: 'dict'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/jwd02f/main/065/717/65717642/configs/tmpxnq1iw99", line 11, in <module>
    adata = ad.read('/data/dnb09/galaxy_db/files/b/e/a/dataset_bea25832-2b41-4b14-8c17-91383d1812bf.dat')
  File "/usr/local/tools/_conda/envs/mulled-v1-f1891eb62511084a03c08c08b82a6b785abdf465b18ed9ac4a36c19217d84a96/lib/python3.9/site-packages/anndata/_io/h5ad.py", line 426, in read_h5ad
    d[k] = read_attribute(f[k])
  File "/usr/local/tools/_conda/envs/mulled-v1-f1891eb62511084a03c08c08b82a6b785abdf465b18ed9ac4a36c19217d84a96/lib/python3.9/functools.py", line 877, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/usr/local/tools/_conda/envs/mulled-v1-f1891eb62511084a03c08c08b82a6b785abdf465b18ed9ac4a36c19217d84a96/lib/python3.9/site-packages/anndata/_io/utils.py", line 162, in func_wrapper
    raise AnnDataReadError(
anndata._io.utils.AnnDataReadError: Above error raised while reading key '/layers' of type <class 'h5py._hl.group.Group'> from /.`

That was for extracting observations but variables and general information also failed. Here is [the history](https://singlecell.usegalaxy.eu/u/j.jakiela/h/seurat---anndata) if you need it. I'm pinging you just in case you have any idea how to fix SCEasy again to avoid this issue.

mtekman commented:
@wee-snufkin  I spoke with @nomadscientist and the error seems to only persist on the singlecell portal because the AnnData tools there are a bit outdated (v0.7.5)

![Screenshot 2023-12-18 at 10-46-02 Galaxy](https://github.com/galaxyproject/training-material/assets/20641402/b9a385b4-edad-4a11-b245-bc54da98e7dc)

I ran the same Inspect tool on usegalaxy.eu but with the modern update from @pavanvidem and everything just worked for that same dataset (v0.10.3):

![Screenshot 2023-12-18 at 10-45-41 Galaxy](https://github.com/galaxyproject/training-material/assets/20641402/cb7fb57e-fdc6-4610-b963-2fe39a00d154)

So I guess we just need to update the tools on singlecell.usegalaxy.eu

mtekman commented:
Oh, I think Pavan already updated the tools!

@wee-snufkin  You just need to use a newer version of the inspect tool (v0.10.x)

pavanvidem commented:
There is an incompatibility of anndata between versions before 0.8 and after 0.8. Please use either everything before v0.8 or everything after v0.8.

hexylena commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,850 @@
+- **Tabular** - simply using TSV, CSV or TXT formats to store expression matrix as well as cell and gene metadata. 
+- **MTX** - it's just a sparse matrix format with genes on the rows and cells on the columns as output by Cell Ranger.
+- **HDF5** - Hierarchical Data Format - can store datasets and groups. A dataset is a  a multidimensional array of data elements, together with supporting metadata. A group is a structure for organizing objects in an HDF5 file. This format allows for storing both the count matrices and all metadata in a single file rather than having separate features, barcodes and matrix files. 
+- **AnnData objects** - [anndata](link https://anndata.readthedocs.io/en/latest/) is a Python package for handling annotated data matrices. In Galaxy, you'll see AnnData objects in **h5ad** format, which is based on the standard HDF5 (h5) format. There are lots of Python tools that work with this format, such as Scanpy, MUON, Cell Oracle, SquidPy, etc. 
```
```suggestion
- **AnnData objects** - [anndata](https://anndata.readthedocs.io/en/latest/) is a Python package for handling annotated data matrices. In Galaxy, you'll see AnnData objects in **h5ad** format, which is based on the standard HDF5 (h5) format. There are lots of Python tools that work with this format, such as Scanpy, MUON, Cell Oracle, SquidPy, etc. 
```

hexylena commented on topics/single-cell/tutorials/scrna-case_monocle3-trajectories/tutorial.md:
```diff
@@ -68,14 +69,16 @@ In this tutorial we will perform trajectory analysis using [monocle3](https://co
-Monocle3 works great with annotated data, so we will make use of our annotated AnnData object, generated in the previous [tutorial]({% link topics/single-cell/tutorials/scrna-case_basic-pipeline/tutorial.md %}). So you see - all the hard work of processing data was not in vain! We will also need a ‘clean’ expression matrix, extracted from the AnnData object just before we started the processing.
-You have two options for uploading these datasets. Importing via history is often faster.
+
+In the previous tutorials, we first created an AnnData object and performed downstream analysis on that file. However, Monocle3 uses another datatype which is Cell Data Set (CDS). To be able to infer trajectories in Monocle, we need to transform our AnnData object into CDS file. And guess what - we already have a tutorial for that! We did it in [format conversion tutorial]({% link topics/single-c ...
```
```suggestion
In the previous tutorials, we first created an AnnData object and performed downstream analysis on that file. However, Monocle3 uses another datatype which is Cell Data Set (CDS). To be able to infer trajectories in Monocle, we need to transform our AnnData object into CDS file. And guess what - we already have a tutorial for that! We did it in [format conversion tutorial]({% link topics/single-cell/tutorials/scrna-data-ingest/tutorial.md %}), in the [Anndata -> Cell Data Set (CDS) subsection]({% link topics/single-cell/tutorials/scrna-data-ingest/tutorial.md %}#anndata---cell-data-set-cds). To better understand the structure of CDS object and learn how to create it from expression matrix, cell and gene annotations, it is highly recommended that you complete the mentioned tutorial before importing the prepared CDS file. 
```

hexylena commented on topics/single-cell/tutorials/scrna-case_monocle3-trajectories/tutorial.md:
```diff
@@ -90,288 +93,49 @@ You have two options for uploading these datasets. Importing via history is ofte
-> > ![Cell annotations along the top, n_genes, n_counts, louvain, cell_type with a cell barcode and subsequent metadata as each row](../../images/scrna-casestudy-monocle/example_cell_annotations.png "Example cell annotations")
-> >
-> {: .solution}
+> Monocle3 turns the expression matrix, cell and gene annotations into an object called cell_data_set (CDS), which holds single-cell expression data. We provided you with the CDS file already, but it was created by combining the three mentioned elements. Check out [data conversion tutorial]({% link topics/single-cell/tutorials/scrna-data-ingest/tutorial.md#anndata---cell-data-set-cds %}) to see h ...
```
```suggestion
> Monocle3 turns the expression matrix, cell and gene annotations into an object called cell_data_set (CDS), which holds single-cell expression data. We provided you with the CDS file already, but it was created by combining the three mentioned elements. Check out [data conversion tutorial]({% link topics/single-cell/tutorials/scrna-data-ingest/tutorial.md %}#anndata---cell-data-set-cds) to see how to do it!
```

wee-snufkin replied:
thanks!

hexylena reviewed (commented):
These changes should fix the linting :)

wee-snufkin commented:
> There is an incompatibility of anndata between versions before 0.8 and after 0.8. Please use either everything before v0.8 or everything after v0.8.

Ok, thanks! How about AnnData Operations and Scanpy FilterCells tools? They both keep on failing on the converted object, no matter the version...

mtekman commented:
> > There is an incompatibility of anndata between versions before 0.8 and after 0.8. Please use either everything before v0.8 or everything after v0.8.
> 
> Ok, thanks! How about AnnData Operations and Scanpy FilterCells tools? They both keep on failing on the converted object, no matter the version...

Filter seems to work for me

![image](https://github.com/galaxyproject/training-material/assets/20641402/1e9741f6-6fee-4436-bd87-daa368b84531)


AnnData Operations I couldn't test because I don't know what it does :grin :

wee-snufkin commented:
> > > Filter seems to work for me
> 
> AnnData Operations I couldn't test because I don't know what it does :grin :

I actually meant [this Scanpy FilterCells tool](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/ebi-gxa/scanpy_filter_cells/scanpy_filter_cells/1.8.1+galaxy93) which I used as an alternative to AnnData Operations to flag mito genes (I wanted to use AnnData Operations in the first place to change filed names in var and flag mito genes)

hexylena commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,871 @@
+  - data import
+---
+
+# Introduction 
```
```suggestion
```

hexylena commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
````diff
@@ -0,0 +1,871 @@
+> 1. Create a new history for this section *"Downsampling FASTQ Files"*
+> 2. Import the files from [Zenodo]({{ page.zenodo_link }})
+>
+>      ```
+>      https://zenodo.org/record/4574153/files/SLX-7632.TAAGGCGA.N701.s_1.r_1.fq-400k.fastq
+>      https://zenodo.org/record/4574153/files/SLX-7632.TAAGGCGA.N701.s_1.r_2.fq-400k.fastq
+>      ```
````
```suggestion
>    ```
>    https://zenodo.org/record/4574153/files/SLX-7632.TAAGGCGA.N701.s_1.r_1.fq-400k.fastq
>    https://zenodo.org/record/4574153/files/SLX-7632.TAAGGCGA.N701.s_1.r_2.fq-400k.fastq
>    ```
```

I suspect this will fix the linting error (I've seen it now twice on two PRs, where it has not once been triggered since we added that lint, which is, odd.)

wee-snufkin replied:
amazing, thanks for doing your magic <3

pavanvidem commented:
> > > > Filter seems to work for me
> > 
> > 
> > AnnData Operations I couldn't test because I don't know what it does :grin :
> 
> I actually meant [this Scanpy FilterCells tool](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/ebi-gxa/scanpy_filter_cells/scanpy_filter_cells/1.8.1+galaxy93) which I used as an alternative to AnnData Operations to flag mito genes (I wanted to use AnnData Operations in the first place to change filed names in var and flag mito genes)

@wee-snufkin Do you have a history with the failing tools? I guess the problem is that the EBI tools still use the old version of Scanpy. Maybe an old scanpy tool somehow worked on the latest version of anndata and produced an older version of anndata. I know it It's confusing but it is just a guess. Updating every scanpy and anndata based tools to the latest versions hopefully fixes the issue.

wee-snufkin commented:
> @wee-snufkin Do you have a history with the failing tools? I guess the problem is that the EBI tools still use the old version of Scanpy. Maybe an old scanpy tool somehow worked on the latest version of anndata and produced an older version of anndata. I know it It's confusing but it is just a guess. Updating every scanpy and anndata based tools to the latest versions hopefully fixes the issue.

Here is the [history](https://usegalaxy.eu/u/j.jakiela/h/seurat---anndata). If I'm not mistaken, the newest version of _AnnData Operations_ uses scanpy 1.1.6, but feel free to check the error in the history

pavanvidem commented:
The latest SCEasy converter converts into the latest version of anndata object. Then it is not possible to use it with any of the EBI SC tools because they support older versions of the anndata. The easiest solution is to use the SCEasy convert from EBI tools that generate anndata compatible with the EBI single-cell tools.

wee-snufkin commented:
> The latest SCEasy converter converts into the latest version of anndata object. Then it is not possible to use it with any of the EBI SC tools because they support older versions of the anndata. The easiest solution is to use the SCEasy convert from EBI tools that generate anndata compatible with the EBI single-cell tools.

@pavanvidem got it, many thanks! Now all works perfectly and I included some explanation for the users regarding the use of the two SCEasy tools.

shiltemann commented:
Hi @wee-snufkin, thanks for your tutorial, this looks all good from a technical point of view, so please go ahead and merge when you are happy with it @nomadscientist or @pavanvidem :)

@wee-snufkin: if you would like to write a news post announcing your new tutorials or other updates, please feel free to always include that in the PR as well (but up to you to decide if that makes sense)

pavanvidem commented:
I still have to test this. I will try to allocate some time today.

wee-snufkin commented:
> @wee-snufkin: if you would like to write a news post announcing your new tutorials or other updates, please feel free to always include that in the PR as well (but up to you to decide if that makes sense)

Thanks @shiltemann! I've added a wee news post to this PR and updated some of the contributions to tutorials :)

pavanvidem commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,881 @@
+>
+{: .hands_on}
+
+Finally, let's combine those files that we have just generated and turn them into the SingleCellExperiment!
```
Alternatively, can be done with SCEasy. Anndata -> Seurat -> SCE

wee-snufkin replied:
Right, will add an additional hands-on box!

pavanvidem commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,881 @@
+>    
+{: .hands_on}
+
+And that's it! The latest tool {% tool [SCEasy Converter](toolshed.g2.bx.psu.edu/repos/iuc/sceasy_convert/sceasy_convert/0.0.7+galaxy2) %} will do the same but the output file will be the newest AnnData version and will not work with the tool used below.  
```
Maybe good to have an alternative hands-on with the latest SCEasy converter tool too.

wee-snufkin replied:
I just mentioned the latest SCEasy here without a hands-on box because I thought it's similar enough to the old one that the users will know how to use it. And it would produce the AnnData output which wouldn't be compatible with the Filter, Plot, Explore workflow and in this section the main aim was to do the conversion and pass on the converted object into that workflow. As you pointed out previously, we need an older version of AnnData for that workflow, hence there it is the older version of SCEasy used here.

pavanvidem commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,881 @@
+
+> <hands-on-title> Modify AnnData object to make it compatible with Filter, Plot, Explore workflow </hands-on-title>
+>
+> 1. {% tool [AnnData Operations](toolshed.g2.bx.psu.edu/repos/ebi-gxa/anndata_ops/anndata_ops/1.8.1+galaxy92) %}
```
Did it work for you? It failed for me.

wee-snufkin replied:
Oh no! 
It did work for me: https://singlecell.usegalaxy.eu/u/j.jakiela/h/seurat---anndata - see dataset 266 old SCEasy conversion, then 268 Anndata operations and all datasets that come later are the outputs of the Filter, Plot, Explore workflow.

pavanvidem reviewed (commented):
Hi, this tutorial is very much needed in the community. We are seeing now and then questions about getting data into Galaxy and converting. A very good start!! 
I liked the concept of this tutorial but it would be nice to have some more alternative hands-ons for each type of conversion. It seems like the tutorial is built to work with other tutorials based on the EBI tools. Maybe some users just need to perform basic clustering with PBMC tutorial. Some of the conversions can be done with the latest SCEasy with a single click to produce the latest version of anndata that is compatible with the PBMC tutorial. 
Currently, it seems like these are the only options to convert. Do you think it makes sense to add an alternative hands-on with SCEasy converter (with choose your tutorial)? Then we can produce the latest and old versions of anndata.

wee-snufkin commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,881 @@
+> 3. Alternatively, you can import history where we created the Seurat object: [Input history](https://singlecell.usegalaxy.eu/u/j.jakiela/h/ebi-scxa---anndata-scanpy-or-seurat-object-1) 
+>
+>    {% snippet faqs/galaxy/histories_import.md %}
+>
```
Thanks for spotting that - I can add a note for the users to change the datatype to rdata when importing a file in this way.

wee-snufkin commented on topics/single-cell/tutorials/scrna-data-ingest/tutorial.md:
```diff
@@ -0,0 +1,881 @@
+
+Most of our conversions involve extracting tables from different data objects and importing them into the target object.
+
+First, we will extract observations (cell metadata) and the full matrix from our AnnData.
```
Indeed, the idea was to show the manual conversion to give the users an idea of the structure of those objects and how they are related. 
The new SCEasy converter was introduced first in this tutorial, because it's the easiest to use and probably does all the conversion the users might need. I thought it's straightforward enough to mention it at the beginning and then just show other methods, but your comment gave me an idea of adding an additional hands-on box here that will show the one-step conversion with the new SCEasy converter.

wee-snufkin commented:
Apologies for keeping updating this PR for so long, but I think I'm done! 
All the tests pass the tutorial part (5debf55), linting only fails for the cyoa bit.

hexylena reviewed (approved):
That linting error is fine :+1:
